create_account_ban.lua

```diff
@@ -0,0 +1,185 @@
+local MSG_BLOCKED = "Contact us on Discord or IRC if this seems wrong: New accounts are blocked from your address."
+local LOG_BLOCKED = "blocked new account name %s from %s %s"
+
+local has_beerchat = minetest.get_modpath("beerchat")
+
+local storage = minetest.get_mod_storage()
+
+local blacklist_ip
+local blacklist_asn
+local active
+local logging
+
+local function inform_admin(msg)
+	minetest.log("action", "[beowulf] " .. msg)
+	if has_beerchat then
+		beerchat.send_on_channel(CHAT_NAME, beerchat.moderator_channel_name, msg)
+	end
+end
+
+local function cleanup()
+	local dirty = false
+	local now = os.time()
+	for key, options in pairs(blacklist_ip) do
+		if options.expire and options.expire >= now then
+			blacklist_ip[key] = nil
+			dirty = true
+		end
+	end
+	for key, options in pairs(blacklist_asn) do
+		if options.expire and options.expire >= now then
+			blacklist_asn[key] = nil
+			dirty = true
+		end
+	end
+	return dirty
+end
+
+local function read_storage()
+	blacklist_ip = minetest.deserialize(storage:get("create_account_ban.blacklist_ip")) or {}
+	blacklist_asn = minetest.deserialize(storage:get("create_account_ban.blacklist_asn")) or {}
+	active = storage:get("create_account_ban.active") ~= nil
+	logging = storage:get_int("create_account_ban.logging")
+end
+
+local function write_storage()
+	storage:set_string("create_account_ban.blacklist_ip", blacklist_ip and minetest.serialize(blacklist_ip) or "")
+	storage:set_string("create_account_ban.blacklist_asn", blacklist_asn and minetest.serialize(blacklist_asn) or "")
+	storage:set_string("create_account_ban.active", active and "1" or "")
+	storage:set_int("create_account_ban.logging", logging)
+end
+
+local function match_values(a, b, options)
+	if type(options) == "table" and options.pattern then
+		return a:find(b) ~= nil
+	end
+	return a == b
+end
+
+minetest.register_on_prejoinplayer(function(name, ip)
+	if active and blacklist_ip and not minetest.get_auth_handler().get_auth(name) then
+		for ip_pattern, options in pairs(blacklist_ip) do
+			if match_values(ip, ip_pattern, options) then
+				-- Account creation blocked by blacklisted IP address
+				if has_beerchat then
+					local msg = "Blocked new account " .. name .. " from " .. ip
+					beerchat.send_on_channel("*Beowulf*", beerchat.moderator_channel_name, msg)
+				end
+				return MSG_BLOCKED
+			end
+		end
+	end
+end)
+
+local function kick_and_delete_player_account(name)
+	minetest.kick_player(name, MSG_BLOCKED)
+	local auth_handler = minetest.get_auth_handler()
+	if auth_handler.get_auth(name) then
+		minetest.log("action", "[beowulf] Removing player account: "..name)
+		auth_handler.delete_auth(name)
+		-- Temporary IP ban could be added here but probably not that useful, can be used to skip temp account creation
+	else
+		-- Should not happen, barrier is most probably leaking for some reason if this ever gets logged
+		minetest.log("error", "[beowulf] Player account should exist but not found: "..name)
+	end
+end
```

Comment:
@BuckarooBanzay / @OgelGames Do you think this is safe enough or is there something that should be changed to make it better?

It is called when new accounts are blocked based on IP/ASN, this is workaround to allow enforcing new account ban with async geoip lookup.

Automated tests would be good (and I'll include some) to prevent account deletion because of some stupid mistake but besides that for actual implementation?

Reply:
> @BuckarooBanzay / @OgelGames (or anyone else) Do you think this is safe enough or is there something that should be changed to make it better?

LGTM, the kick message should be clear enough i think :+1: